models/user.go

```diff
@@ -1402,35 +1406,144 @@ func SyncExternalUsers() {
 					}
 
 					err = CreateUser(usr)
+
 					if err != nil {
 						log.Error(4, "SyncExternalUsers[%s]: Error creating user %s: %v", s.Name, su.Username, err)
+					} else if isAttributeSSHPublicKeySet {
+						// Add Public SSH Keys to User from LDAP
+						for _, LDAPPublicSSHKey := range su.SSHPublicKey {
+							log.Trace("SyncExternalUsers[%s]: Adding LDAP Public SSH Key for user %s: %v", s.Name, usr.Name, LDAPPublicSSHKey)
+							LDAPPublicSSHKeyName := strings.Join([]string{s.Name, LDAPPublicSSHKey[0:40]}, "-")
+							key, err := AddPublicKey(usr.ID, LDAPPublicSSHKeyName, LDAPPublicSSHKey, s.ID)
+							if err != nil {
+								log.Error(4, "SyncExternalUsers[%s]: Error adding LDAP Public SSH Key for user %s: %v [%v]", s.Name, su.Username, key, err)
+							} else {
+								sshKeysNeedUpdate = true
+							}
+						}
 					}
 				} else if updateExisting {
 					existingUsers = append(existingUsers, usr.ID)
-					// Check if user data has changed
-					if (len(s.LDAP().AdminFilter) > 0 && usr.IsAdmin != su.IsAdmin) ||
-						strings.ToLower(usr.Email) != strings.ToLower(su.Mail) ||
-						usr.FullName != fullName ||
-						!usr.IsActive {
-
-						log.Trace("SyncExternalUsers[%s]: Updating user %s", s.Name, usr.Name)
-
-						usr.FullName = fullName
-						usr.Email = su.Mail
-						// Change existing admin flag only if AdminFilter option is set
-						if len(s.LDAP().AdminFilter) > 0 {
-							usr.IsAdmin = su.IsAdmin
-						}
-						usr.IsActive = true
 
-						err = UpdateUser(usr)
+					if isAttributeSSHPublicKeySet {
+						log.Trace("SyncExternalUsers[%s]: Handling user %s", s.Name, usr.Name)
+
+						// Get Public Keys from DB with LDAP source
+						var giteaKeys []string
+						keys, err := ListPublicKeys(usr.ID)
+
 						if err != nil {
-							log.Error(4, "SyncExternalUsers[%s]: Error updating user %s: %v", s.Name, usr.Name, err)
+							log.Error(4, "SyncExternalUsers[%s]: Error listing LDAP Public SSH Keys for user %s: %v", s.Name, usr.Name, err)
+						}
+						for _, v := range keys {
+							// If key was synced from LDAP
+							if v.LoginSourceID > 0 {
+								giteaKeys = append(giteaKeys, v.OmitEmail())
+							}
+						}
+
+						// Get Public Keys from LDAP and skip duplicate keys
+						var ldapKeys []string
+						for _, v := range su.SSHPublicKey {
+							ldapKey := strings.Join(strings.Split(v, " ")[:2], " ")
+							// Check if value is already in slice, else add it
+							i := sort.SearchStrings(ldapKeys, ldapKey)
+							if i < len(ldapKeys) && ldapKeys[i] == ldapKey {
+							} else {
+								ldapKeys = append(ldapKeys, ldapKey)
+							}
+						}
+
+						// Sort key slices
+						sort.Strings(ldapKeys)
+						sort.Strings(giteaKeys)
+
+						// Check if sync is needed
+						var giteaKeysToDelete []string
+						if reflect.DeepEqual(giteaKeys, ldapKeys) {
```

Comment:
giteaKeys and ldapKeys are of the same types, slice of strings.

If I were to perform a compare between giteaKeys and ldapKeys in the suggested way, I'd need to create PublicKey types instead of string slices. While this is doable, it would require a fingerprint calculation for every LDAP key as the keys are stored as strings (content only) in LDAP (this is usually how sysadmins choose to store them). In other words; the fingerprint is only stored in database (calcFingerprint is called from addKey).

This also seems most convenient as the addKey function takes a string for content-value (content string).

As both the database and LDAP contains the key content, it seems more appropriate to just compare them.

As I'm building the key name out of the key string (LDAPPublicSSHKey[0:40]) it doesn't matter if the key name (description) for the key has been updated in LDAP.

We're therefore only interested in checking if the actual content has changed.

I've created a new function to compare the slices (giteaKeys and ldapKeys), which is more efficient than reflect.DeepEqual.